src/pybind/mgr/dashboard/tools.py

```diff
@@ -47,7 +48,7 @@ def request_begin(self):
             msg = '[DASHBOARD] from=\'{}\' path=\'{}\' method=\'{}\' ' \
                 'user=\'{}\''.format(url, req.path_info, req.method, user)
             if Settings.AUDIT_API_LOG_PAYLOAD:
-                params = req.params if req.params else {}
+                params = copy.deepcopy(req.params) if req.params else {}
                 params.update(get_request_body_params(req))
```

Comment:
In this case, something like this might also work:
```python
params = dict(req.params or {}, **get_request_body_params(req))
```
As long as you don't modify objects (like `update()`), you typically don't need copy.deepcopy

Reply:
The params are modified in a [later step](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/tools.py#L62), so i think a deepcopy is required.

Reply:
Saw that already. You're fine, as long as you don't modify the values, like so:
```python
params[key].append('***')
``` 
This is, because you are only modifying `params`

Reply:
Turns out, even this is safe:
```python
In [1]: d1 = {'1': 'a', '2': 'b'}
In [2]: d2 = {'3': 'c', '4': 'd'}
In [3]: d3 = dict(d1, **d2)

In [4]: d3
Out[4]: {'1': 'a', '2': 'b', '3': 'c', '4': 'd'}

In [5]: d3['1'] = '***'
In [6]: d3, d1
Out[6]: ({'1': '***', '2': 'b', '3': 'c', '4': 'd'}, {'1': 'a', '2': 'b'})

In [9]: d3['3'] += 'WRONG'
In [10]: d3, d2
Out[10]: ({'1': '***', '2': 'b', '3': 'cWRONG', '4': 'd'}, {'3': 'c', '4': 'd'})

```

Reply:
Done